File: 11-container-with-most-water/container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        // formula int minH = min(height[L], height[R]);
        // difference int R-L = 7 ;
        // maxA = max(maxA, diff*minH);
        int maxA = INT_MIN;
        int L = 0;
        int R = height.size() - 1;
        while (L < R) {
            int minH = min(height[L], height[R]);
            int diff = R - L;
            int area = minH * diff;
            maxA = max(maxA, area);
            if (height[L] < height[R]) {
                L++;
            } else {
                R--;
            }
        }
        return maxA;
    }
};

File: 11-container-with-most-water/container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
    //  find the max number and then try to find the second maximum and so on and compare the difference we can get our results its O(n) solution.
    // 2nd solution, two pointer why because of the search space we need to search something
    // move to smaller one untill L < R
    let indexed = height.map((h, i) => [h, i]);

    // sort descending by height
    indexed.sort((a, b) => b[0] - a[0]);

    let max = 0;
    let minIndex = indexed[0][1];
    let maxIndex = indexed[0][1];

    for (let i = 1; i < indexed.length; i++) {
        let [h, idx] = indexed[i];

        let width = Math.max(Math.abs(idx - minIndex), Math.abs(idx - maxIndex));
        max = Math.max(max, width * h);

        minIndex = Math.min(minIndex, idx);
        maxIndex = Math.max(maxIndex, idx);
    }

    return max;
};